File: fsharp/alphabet-cipher/test/coder-test.fs
module coder_test

open Expecto
open alphabet_cipher.coder

[<Tests>]
let tests =
    testList "coder" [
        testCase "can encode a message with a secret keyword" <| fun _ ->
            Expect.equal
                (encode "vigilance" "meetmeontuesdayeveningatseven") "hmkbxebpxpmyllyrxiiqtoltfgzzv" ""
            Expect.equal
                (encode "scones" "meetmebythetree") "egsgqwtahuiljgs" ""

        testCase "can decode a message given an encoded message and a secret keyword" <| fun _ ->
            Expect.equal
                (decode "vigilance" "hmkbxebpxpmyllyrxiiqtoltfgzzv") "meetmeontuesdayeveningatseven" ""
            Expect.equal
                (decode "scones" "egsgqwtahuiljgs") "meetmebythetree" ""

        testCase "can extract the secret keyword given an encrypted message and the original message" <| fun _ ->
            Expect.equal
                (decipher "opkyfipmfmwcvqoklyhxywgeecpvhelzg" "thequickbrownfoxjumpsoveralazydog") "vigilance" ""
            Expect.equal
                (decipher "hcqxqqtqljmlzhwiivgbsapaiwcenmyu" "packmyboxwithfivedozenliquorjugs") "scones" ""
            Expect.equal
                (decipher "hfnlphoontutufa" "hellofromrussia") "abcabcx" ""
    ]

[<EntryPoint>]
let main argv =
    runTestsInAssembly defaultConfig argv

File: csharp/alphabet-cipher/test/coder-test.cs
namespace alphabet_cipher;

using Xunit;
using static alphabet_cipher.coder;

public class coder_test {
    [Fact]
    public void encode_message_with_keyword() {
        Assert.Equal (
            "hmkbxebpxpmyllyrxiiqtoltfgzzv", encode("vigilance", "meetmeontuesdayeveningatseven")
        );
        Assert.Equal (
            "egsgqwtahuiljgs", encode("scones", "meetmebythetree")
        );
    }

    [Fact]
    public void decode_encoded_message_with_keyword() {
        Assert.Equal (
            "meetmeontuesdayeveningatseven", decode("vigilance", "hmkbxebpxpmyllyrxiiqtoltfgzzv")
        );
        Assert.Equal (
            "meetmebythetree", decode("scones", "egsgqwtahuiljgs")
        );
    }

    [Fact]
    public void extract_keyword_from_encrypted_message_and_message() {
        Assert.Equal (
            "vigilance", decipher("opkyfipmfmwcvqoklyhxywgeecpvhelzg", "thequickbrownfoxjumpsoveralazydog")
        );
        Assert.Equal (
            "scones", decipher("hcqxqqtqljmlzhwiivgbsapaiwcenmyu", "packmyboxwithfivedozenliquorjugs")
        );
        Assert.Equal (
            "abcabcx", decipher("hfnlphoontutufa", "hellofromrussia")
        );
    }
}
